Bottom: 7380d79b1528a3cb9a0935e73577b46ca849cb69
Top:    ac4e3981d633544366582ee92e2449affd412d16
Author: Johannes Rudolph <johannes_rudolph@gmx.de>
Date:   2009-09-29 16:36:45 +0200

binary function generation support


---

diff --git a/bytecode/src/main/scala/net/virtualvoid/bytecode/Bytecode.scala b/bytecode/src/main/scala/net/virtualvoid/bytecode/Bytecode.scala
index 5028d8a..6f9fe6e 100644
--- a/bytecode/src/main/scala/net/virtualvoid/bytecode/Bytecode.scala
+++ b/bytecode/src/main/scala/net/virtualvoid/bytecode/Bytecode.scala
@@ -268,6 +268,9 @@ object Bytecode{
 	  def compile[T<:AnyRef,U<:AnyRef](cl:Class[T])(
                        code: Local[T] => F[Nil] => F[Nil**U]   
 	  ): T => U
+	  def compile[T1<:AnyRef,T2<:AnyRef,U<:AnyRef](cl1:Class[T1],cl2:Class[T2])(
+	    code: (Local[T1],Local[T2]) => F[Nil] => F[Nil**U]
+	  ): (T1,T2) => U
   }
 
   trait RichFunc[ST1<:List,ST2<:List] 
diff --git a/bytecode/src/main/scala/net/virtualvoid/bytecode/Compiler.scala b/bytecode/src/main/scala/net/virtualvoid/bytecode/Compiler.scala
index 9537d92..5c9aab5 100644
--- a/bytecode/src/main/scala/net/virtualvoid/bytecode/Compiler.scala
+++ b/bytecode/src/main/scala/net/virtualvoid/bytecode/Compiler.scala
@@ -298,6 +298,13 @@ object ASMCompiler extends ByteletCompiler{
         code(local(1,cl))(new ASMFrame[Nil](mv,EmptyClassStack,2))
       }
     }
+    def compile[T1<:AnyRef,T2<:AnyRef,U<:AnyRef](cl1:Class[T1],cl2:Class[T2])(
+	    code: (Local[T1],Local[T2]) => F[Nil] => F[Nil**U]
+	  ): (T1,T2) => U = {
+      classStub[(T1,T2)=>U](cl1,cl2){ mv =>
+        code(local(1,cl1),local(2,cl2))(new ASMFrame[Nil](mv,EmptyClassStack,3))
+      }
+    }
     def classStub[T](params:Class[_]*)(body: MethodVisitor => Unit) :T = {
       val numParams = params.length
       val superClass = "net/virtualvoid/bytecode/AbstractFunction"+numParams
@@ -349,3 +356,4 @@ object ASMCompiler extends ByteletCompiler{
   }
 
 abstract class AbstractFunction1[T,U] extends Function1[T,U]
+abstract class AbstractFunction2[T1,T2,U] extends Function2[T1,T2,U]
\ No newline at end of file
diff --git a/bytecode/src/main/scala/net/virtualvoid/bytecode/Interpreter.scala b/bytecode/src/main/scala/net/virtualvoid/bytecode/Interpreter.scala
index b5974aa..9baca87 100644
--- a/bytecode/src/main/scala/net/virtualvoid/bytecode/Interpreter.scala
+++ b/bytecode/src/main/scala/net/virtualvoid/bytecode/Interpreter.scala
@@ -128,4 +128,8 @@ object Interpreter extends ByteletCompiler{
         def load[ST<:List]:F[ST] => F[ST**T] = f => IF(f.stack**value)
         def store[ST<:List]:F[ST**T] => F[ST] = f => {value = f.stack.top; IF(f.stack.rest)}
       })(IF(N)).stack.top
+    
+    def compile[T1<:AnyRef,T2<:AnyRef,U<:AnyRef](cl1:Class[T1],cl2:Class[T2])(
+	    code: (Local[T1],Local[T2]) => F[Nil] => F[Nil**U]
+	  ): (T1,T2) => U = null
   }
\ No newline at end of file
diff --git a/bytecode/src/test/scala/net/virtualvoid/bytecode/BytecodeCompilerSpecs.scala b/bytecode/src/test/scala/net/virtualvoid/bytecode/BytecodeCompilerSpecs.scala
index c3327a4..d7889b9 100644
--- a/bytecode/src/test/scala/net/virtualvoid/bytecode/BytecodeCompilerSpecs.scala
+++ b/bytecode/src/test/scala/net/virtualvoid/bytecode/BytecodeCompilerSpecs.scala
@@ -234,6 +234,14 @@ object BytecodeCompilerSpecs extends Specification{
       f(null) must be_==("isnull")
       f("blub") must be_==("blub isnotnull")
     }
+    "binary methods" in {
+      compiler.compile(classOf[String],classOf[String])((str1,str2) =>
+        _ ~
+          str1.load ~
+          str2.load ~
+          invokemethod2(_.concat(_))
+      )("String1","String2") must be_==("String1String2")
+    }
   }
   def array(els:Int*):Array[Int] = Array(els:_*)
   def array(els:String*):Array[String] = Array(els:_*)
